trl/trainer/kto_trainer.py

```diff
@@ -0,0 +1,1075 @@
+
+        if len(prompt_input_ids) != len(prompt_attention_mask):
+            raise ValueError("Prompt input ids and attention mask should have the same length.")
+
+        answer_input_ids = full_tokenized["input_ids"][response_token_ids_start_idx:]
+        answer_attention_mask = full_tokenized["attention_mask"][response_token_ids_start_idx:]
+
+        return dict(
+            prompt_input_ids=prompt_input_ids,
+            prompt_attention_mask=prompt_attention_mask,
+            input_ids=answer_input_ids,
+            attention_mask=answer_attention_mask,
+        )
+
+    def shuffle_completion(self, batch) -> Dict:
+        batch["kl"] = [False] * len(batch["prompt"]) + [True] * len(batch["prompt"])
+        batch["prompt"] = batch["prompt"] + batch["prompt"]
+        batch["completion"] = batch["completion"] + random.sample(batch["completion"], len(batch["completion"]))
+        batch["label"] = batch["label"] + batch["label"]
+        return batch
+
+    def tokenize_row(self, feature, model: Union[PreTrainedModel, nn.Module] = None) -> Dict:
+        """Tokenize a single row from a KTO specific dataset.
+
+        At this stage, we don't convert to PyTorch tensors yet; we just handle the truncation
+        in case the prompt + completion responses is/are too long. First
+            we truncate the prompt; if we're still too long, we truncate the completion.
+
+        We also create the labels for the completion responses, which are of length equal to
+            the sum of the length of the prompt and the completion response, with
+            label_pad_token_id  for the prompt tokens.
+        """
+        batch = {}
+        prompt = feature["prompt"]
+        completion = feature["completion"]
+
+        if not self.is_encoder_decoder:
+            # Check issues below for more details
+            #  1. https://github.com/huggingface/trl/issues/907
+            #  2. https://github.com/EleutherAI/lm-evaluation-harness/pull/531#issuecomment-1595586257
+            #  3. https://github.com/LianjiaTech/BELLE/issues/337
+
+            if not isinstance(prompt, str):
+                raise ValueError(f"prompt should be an str but got {type(prompt)}")
+            prompt_tokens = self.tokenizer(prompt, add_special_tokens=False)
+            prompt_tokens = {f"prompt_{k}": v for k, v in prompt_tokens.items()}
+
+            if not isinstance(completion, str):
+                raise ValueError(f"completion should be an str but got {type(completion)}")
+            completion_tokens = self.build_tokenized_answer(prompt, completion)
+
+            # Last prompt token might get merged by tokenizer and
+            # it should not be included for generation if that happens
+            completion_prompt_len_input_ids = len(completion_tokens["prompt_input_ids"])
+
+            for k, v in prompt_tokens.items():
+                prompt_tokens[k] = v[:completion_prompt_len_input_ids]
+
+            # add BOS token to head of prompt
```

Comment:
Does this make sense for models with a chat template? E.g. if I use ChatML won't we be adding a BOS token to the start of the dialogue which was never seen during SFT ?

Reply:
@lewtun this is for the enc-decoder model where we add a BOS to the start of prompt for the encoder and there is no such token for the completion going to the decoder

Reply:
my bad... so i add BOS to start of prompt only and EOS to the end of the completion